src/features/codeLensProvider.ts

```diff
@@ -6,171 +6,268 @@
+        if (codeLens instanceof ReferencesCodeLens) {
+            return this.resolveReferencesCodeLens(codeLens, token);
+        }
+        else if (codeLens instanceof RunTestsCodeLens) {
+            return this.resolveTestCodeLens(codeLens, 'Run Test', 'dotnet.test.run', 'Run All Tests', 'dotnet.classTests.run');
+        }
+        else if (codeLens instanceof DebugTestsCodeLens) {
+            return this.resolveTestCodeLens(codeLens, 'Debug Test', 'dotnet.test.debug', 'Debug All Tests', 'dotnet.classTests.debug');
+        }
+    }
+
+    private async resolveReferencesCodeLens(codeLens: ReferencesCodeLens, token: vscode.CancellationToken): Promise<vscode.CodeLens> {
+        const request: protocol.FindUsagesRequest = {
+            FileName: codeLens.fileName,
+            Line: codeLens.range.start.line + 1, // OmniSharp is 1-based
+            Column: codeLens.range.start.character + 1, // OmniSharp is 1-based
+            OnlyThisFile: false,
+            ExcludeDefinition: true
+        };
 
-    private async _convertQuickFix(bucket: vscode.CodeLens[], fileName: string, node: protocol.Node, options: Options): Promise<void> {
+        const result = await serverUtils.findUsages(this._server, request, token);
 
-        if (node.Kind === 'MethodDeclaration' && OmniSharpCodeLensProvider.filteredSymbolNames[node.Location.Text]) {
+        if (!result || !result.QuickFixes) {
             return;
         }
 
-        let lens = new OmniSharpCodeLens(fileName, toRange(node.Location));
-        if (options.showReferencesCodeLens) {
-            bucket.push(lens);
-        }
+        const quickFixes = result.QuickFixes;
+        const count = quickFixes.length;
 
-        for (let child of node.ChildNodes) {
-            this._convertQuickFix(bucket, fileName, child, options);
-        }
+        codeLens.command = {
+            title: count === 1 ? '1 reference' : `${count} references`,
+            command: 'editor.action.showReferences',
+            arguments: [vscode.Uri.file(request.FileName), codeLens.range.start, quickFixes.map(toLocation)]
+        };
 
-        if (options.showTestsCodeLens) {
-            await this._updateCodeLensForTest(bucket, fileName, node);
-        }
+        return codeLens;
     }
 
-    resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken): Thenable<vscode.CodeLens> {
-        if (codeLens instanceof OmniSharpCodeLens) {
-
-            let req = <protocol.FindUsagesRequest>{
-                FileName: codeLens.fileName,
-                Line: codeLens.range.start.line + 1,
-                Column: codeLens.range.start.character + 1,
-                OnlyThisFile: false,
-                ExcludeDefinition: true
+    private async resolveTestCodeLens(codeLens: TestCodeLens, singularTitle: string, singularCommandName: string, pluralTitle: string, pluralCommandName: string): Promise<vscode.CodeLens> {
+        if (codeLens.testMethodNames.length === 1) {
```

Comment:
Does this mean that if we have a class which has a single test in it , we will display - "Run Test" instead of "Run All Tests" ?

Reply:
Yes. That was a mistake. Will fix.